Framework/Geometry/src/Objects/ShapeFactory.cpp

```diff
@@ -1416,5 +1709,92 @@ void ShapeFactory::createGeometryHandler(Poco::XML::Element *pElem, std::shared_
   geomHandler->setShapeInfo(std::move(shapeInfo));
 }
 
+/**
+ * Generates a rotate Matrix applying the x rotate then y rotate, then z
+ * rotate
+ * @param xrotate The x rotate required in radians
+ * @param yrotate The y rotate required in radians
+ * @param zrotate The z rotate required in radians
+ * @returns a matrix of doubles to use as the rotate matrix
+ */
+Kernel::Matrix<double> ShapeFactory::generateMatrix(double xrotate, double yrotate, double zrotate) {
+  Kernel::Matrix<double> xMatrix = generateXRotation(xrotate);
+  Kernel::Matrix<double> yMatrix = generateYRotation(yrotate);
+  Kernel::Matrix<double> zMatrix = generateZRotation(zrotate);
+
+  return zMatrix * yMatrix * xMatrix;
+}
+
+/**
+ *Generates the x component of the rotate matrix
+ *@param xrotate The x rotate required in radians
+ *@returns a matrix of doubles to use as the x axis rotate matrix
+ */
+Kernel::Matrix<double> ShapeFactory::generateXRotation(double xrotate) {
+  const double sinX = sin(xrotate);
+  const double cosX = cos(xrotate);
+  std::vector<double> matrixList = {1, 0, 0, 0, cosX, -sinX, 0, sinX, cosX};
+  return Kernel::Matrix<double>(matrixList);
+}
+
+/**
+ * Generates the y component of the rotate matrix
+ * @param yrotate The y rotate required in radians
+ * @returns a matrix of doubles to use as the y axis rotate matrix
+ */
+Kernel::Matrix<double> ShapeFactory::generateYRotation(double yrotate) {
+  const double sinY = sin(yrotate);
+  const double cosY = cos(yrotate);
+  std::vector<double> matrixList = {cosY, 0, sinY, 0, 1, 0, -sinY, 0, cosY};
+  return Kernel::Matrix<double>(matrixList);
+}
+
+/**
+ * Generates the z component of the rotate matrix
+ * @param zrotate The z rotate required in radians
+ * @returns a matrix of doubles to use as the z axis rotate matrix
+ */
+Kernel::Matrix<double> ShapeFactory::generateZRotation(double zrotate) {
+  const double sinZ = sin(zrotate);
+  const double cosZ = cos(zrotate);
+  std::vector<double> matrixList = {cosZ, -sinZ, 0, sinZ, cosZ, 0, 0, 0, 1};
+  return Kernel::Matrix<double>(matrixList);
+}
```

Comment:
The 4 methods above (generateMatrix, generateXRotation, generateYRotation, generateZRotation) feel like they should be done using some pre-existing Mantid code rather than writing your own methods to achieve this. The way Mantid manages rotations is using the Quat class (short for Quaternion). I think you can rewrite generateMatrix to use the Quat class instead.

Unfortunately the Quaternion concept is a bit dense so you may need a lie down first! Fortunately the Goniometer::recalculateR method is some code which is pretty similar to what you want. It loops through the goniometer axes creating a Quat for each and then multiplies them together. Your code needs to do something similar by looping through the 3 x, y, z axes

It's probably worth getting the unit tests to where you want them to be before doing this change so you can be sure it doesn't break anything.

Reply:
I did conveniently ignore the Quat rotation as it seems quite confusing, but happy to have a go (as you rightly said) after I get some unit tests working. I did get these 4 methods from Mantid, I copied them across from MeshFileIO.cpp.  Otherwise I would have had to give Framework/Geometry access to all of Framework/DataHandling to avoid this copy.